transports/tcp/src/lib.rs

```diff
@@ -22,362 +22,597 @@
+        let me = Pin::into_inner(self);
 
-#[cfg(feature = "async-std")]
-impl AsyncRead for TcpTransStream {
-    fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<Result<usize, io::Error>> {
-        AsyncRead::poll_read(Pin::new(&mut self.inner), cx, buf)
-    }
-}
-
-#[cfg(feature = "async-std")]
-impl AsyncWrite for TcpTransStream {
-    fn poll_write(mut self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize, io::Error>> {
-        AsyncWrite::poll_write(Pin::new(&mut self.inner), cx, buf)
-    }
+        loop {
+            match &mut me.in_addr {
+                InAddr::Any { if_watch, addrs } => match if_watch {
+                    // If we listen on all interfaces, wait for `if-watch` to be ready.
+                    IfWatch::Pending(f) => match ready!(Pin::new(f).poll(cx)) {
+                        Ok(w) => {
+                            *if_watch = IfWatch::Ready(w);
+                            continue
+                        }
+                        Err(e) => {
+                            me.disable_port_reuse();
+                            return Poll::Ready(Some(Err(e)))
+                        }
+                    },
+                    // Consume all events for up/down interface changes.
+                    IfWatch::Ready(watch) => while let Some(ev) = watch.next().now_or_never() {
+                        match ev {
+                            Ok(IfEvent::Up(inet)) => {
+                                let ip = inet.addr();
+                                if me.listen_addr.is_ipv4() == ip.is_ipv4() {
+                                    if addrs.insert(ip) {
+                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                                        log::debug!("New listen address: {}", ma);
+                                        me.port_reuse.register(ip, me.listen_addr.port());
+                                        return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(ma))));
+                                    }
+                                }
+                            }
+                            Ok(IfEvent::Down(inet)) => {
+                                let ip = inet.addr();
+                                if me.listen_addr.is_ipv4() == ip.is_ipv4() {
+                                    if addrs.remove(&ip) {
+                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                                        log::debug!("Expired listen address: {}", ma);
+                                        me.port_reuse.unregister(ip, me.listen_addr.port());
+                                        return Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(ma))));
+                                    }
+                                }
+                            }
+                            Err(e) => {
+                                me.disable_port_reuse();
```

Comment:
why disable port reuse if if watch returns an error? it should be just logged and ignored

Reply:
The scheme here is that the addresses of the `TcpListenStream` are unregistered from port reuse whenever the listener encounters a fatal error or is dropped. Here in `poll_next()`, whenever we return `Poll::Ready(Some(Err(e)))` that is a fatal error as per the contract of the `Transport::Listener` (in contrast to `Poll::Ready(Some(Ok(ListenerEvent::Error(e))))`). In this particular case I wasn't sure how fatal errors from `if-watch` are, so I chose to be conservative. If you think `if-watch` errors shouldn't be fatal for the `TcpListenStream`, we can emit a `ListenerEvent::Error` instead, without calling `disable_port_reuse()`.